elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchSchemaFactory.java

```diff
@@ -53,6 +60,24 @@
 
   private static final Logger LOGGER = LoggerFactory.getLogger(ElasticsearchSchemaFactory.class);
 
+  // RestClient objects allocate system resources and are thread safe. Here, we
+  // cache them using a key derived from the hashCode()s of the parameters that
+  // define a RestClient.
+  private static Cache<Integer, RestClient> restClients = CacheBuilder.newBuilder()
+      .maximumSize(1000)
```

Comment:
@ILuffZhe the choices were just made conservatively (no one will connect to more than 1000 ElasticSearch servers and replacing each RestClient at most hourly will not amount to a great load.  Testing a custom build of Drill based on a custom build Calcite would be quite lot of effort for us so can I propose that a first PR just aims to stop the resource leaking?  We can then monitor it for an extended period in one of our Drill environments and report back to you with our data if it looks like these parameters need fine tuning.

Reply:
Thanks for the clarification. I think this PR is better than the former one(as far as I can remember).
It's better not to force push before the final decision, so other reviewers can take a look.

Reply:
What happens if after one hour the client is removed and closed but there is still an active reference inside the `ElasticsearchSchema`?

Reply:
@zabetak thank you, that is indeed a mistake.  The ElasticsearchSchema must either be made to either only obtain a rest client from the cache (remove its own reference) or the cache expiry should be set to "never" which is the approach in the Cassandra adapter.  Which is preferred?

Reply:
Resolved, somewhat.  Clients can still be removed and closed but only if their total number has reached an excessive number, not because of any configured expiry time.  A warning that trouble has begun is logged.  Note that if the clients were not removed and closed and no warning was logged then trouble would nevertheless have begun because the application was heading the road to getting killed for leaking resources.